onnxruntime/core/providers/cpu/tensor/where_op.cc

```diff
@@ -0,0 +1,195 @@
+        } else {
+          std::fill(output.begin(), output.end(), T{});
+        }
+      },
+      [target](gsl::span<T> output, gsl::span<const bool> condition, const T& value) {
+        std::transform(condition.cbegin(), condition.cend(), output.begin(),
+                       [target, &value](bool condition_element) {
+                         return condition_element == target ? value : T{};
+                       });
+      },
+      [target](gsl::span<T> output, gsl::span<const bool> condition, gsl::span<const T> value) {
+        std::transform(condition.cbegin(), condition.cend(), value.cbegin(), output.begin(),
+                       [target](bool condition_element, const T& value_element) {
+                         return condition_element == target ? value_element : T{};
+                       });
+      });
+}
+
+template <typename T>
+std::unique_ptr<Tensor> Select(bool target, const Tensor& condition_tensor, const Tensor& value_tensor,
+                               TensorAllocator<T>& tensor_allocator) {
+  TBroadcaster<bool, T> select_broadcaster{condition_tensor, value_tensor};
+  std::unique_ptr<Tensor> select_tensor{
+      tensor_allocator.Allocate(select_broadcaster.GetOutputShape())};
+  TBroadcastOutput<T> select_broadcast_output{
+      select_broadcaster.GetSpanSize(), *select_tensor};
+
+  SelectBroadcastLoop(target, &select_broadcaster, &select_broadcast_output);
+
+  return select_tensor;
+}
+
+template <typename T>
+std::enable_if_t<IsEigenScalarCompatible<T>, void>
+MergeBroadcastLoop(TBroadcaster<T, T>* merge_broadcaster, TBroadcastOutput<T>* merge_broadcast_output) {
+  const auto merge_scalar_vector = [](EigenVectorMap<T> output,
+                                      const T& scalar_value, ConstEigenVectorMap<T> vector_value) {
+    if (scalar_value != T{}) {
+      output = EigenVectorMap<T>::PlainObject::Constant(vector_value.size(), scalar_value);
+    } else {
+      output = vector_value;
+    }
+  };
+
+  BroadcastLoop(
+      *merge_broadcaster, *merge_broadcast_output,
+      [merge_scalar_vector](EigenVectorMap<T> output, const T& X_selection, ConstEigenVectorMap<T> Y_selection) {
+        merge_scalar_vector(output, X_selection, Y_selection);
+      },
+      [merge_scalar_vector](EigenVectorMap<T> output, ConstEigenVectorMap<T> X_selection, const T& Y_selection) {
+        merge_scalar_vector(output, Y_selection, X_selection);
+      },
+      [](EigenVectorMap<T> output, ConstEigenVectorMap<T> X_selection, ConstEigenVectorMap<T> Y_selection) {
+        output = X_selection.binaryExpr(Y_selection, [](T x, T y) -> T {
+          return x != T{} ? x : y;
+        });
+      });
+}
+
+template <typename T>
+std::enable_if_t<!IsEigenScalarCompatible<T>, void>
+MergeBroadcastLoop(TBroadcaster<T, T>* merge_broadcaster, TBroadcastOutput<T>* merge_broadcast_output) {
+  const auto merge_scalar_vector = [](gsl::span<T> output, const T& scalar_value, gsl::span<const T> vector_value) {
```

Comment:
>merge_scalar_vector [](start = 13, length = 19)

Is this merging a scalar or non-scalar vector?

Reply:
It's merging a scalar (scalar_value) and a vector (vector_value).

Reply:
Renamed it, hopefully it's clearer now.